pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -2,301 +2,1132 @@ package listusers
+			type user
+			type document
+				relations
+					define owner: [user]
+					define editor: [user] or owner
+					define viewer: owner or editor`,
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "owner", "user:will"),
+				tuple.NewTupleKey("document:1", "editor", "user:maria"),
+				tuple.NewTupleKey("document:2", "viewer", "user:jon"),
+			},
+			expectedUsers: []string{"user:will", "user:maria"},
 		},
 		{
-			name: "direct_relationship_through_userset",
+			name: "computed_relationship_with_contextual_tuples",
 			req: &openfgav1.ListUsersRequest{
-				StoreId:              storeID,
-				AuthorizationModelId: modelID,
-				Object:               &openfgav1.Object{Type: "document", Id: "1"},
-				Relation:             "viewer",
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
 				UserFilters: []*openfgav1.ListUsersFilter{
 					{
 						Type: "user",
 					},
 				},
+				ContextualTuples: &openfgav1.ContextualTupleKeys{
+					TupleKeys: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("document:1", "owner", "user:will"),
+						tuple.NewTupleKey("document:1", "owner", "user:maria"),
+						tuple.NewTupleKey("document:2", "viewer", "user:jon"),
+					},
+				},
 			},
-			typeDefinitions: parser.MustParse(`
+			model: `model
+			schema 1.1
 			type user
-			type group
-			  relations
-			    define member: [user] as self
 			type document
-			  relations
-			    define viewer: [group#member] as self
-			`),
+				relations
+					define owner: [user]
+					define viewer: owner`,
+			tuples:        []*openfgav1.TupleKey{},
+			expectedUsers: []string{"user:will", "user:maria"},
+		},
+	}
+	tests.runListUsersTestCases(t)
+}
+
+func TestListUsersUsersets(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+	model := `model
+	schema 1.1
+	type user
+	type group
+		relations
+			define member: [user]
+	type document
+		relations
+			define viewer: [group#member]`
+
+	tests := ListUsersTests{
+		{
+			name: "userset_user_granularity",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model: model,
 			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("group:eng", "member", "user:will"),
+				tuple.NewTupleKey("group:eng", "member", "user:maria"),
+				tuple.NewTupleKey("group:marketing", "viewer", "user:jon"),
 				tuple.NewTupleKey("document:1", "viewer", "group:eng#member"),
-				tuple.NewTupleKey("group:eng", "member", "user:jon"),
 			},
-			expectedUsers: []*openfgav1.User{
-				{
-					User: &openfgav1.User_Object{
-						Object: &openfgav1.Object{Type: "user", Id: "jon"},
+			expectedUsers: []string{"user:will", "user:maria"},
+		},
+		{
+			name:                  "userset_group_granularity",
+			TemporarilySkipReason: "because `group:eng` is being returned instead of `group:fga#member`",
```

Comment:
Did you mean "because `group:eng` is being returned instead of `group:eng#member`?

I don't see `group:fga` referenced anywhere in this specific test case.

Reply:
Yep. Just a typo.